### PR TITLE
Calculation error fixes

### PR DESCRIPTION
**Pull request by Humonitarian**
## About The Pull Request

Apparently I knew my code wrong, asked SPCR about it and he corrected my error. I looked through most of my previous PRs to find any cases where I made this mistake, and found a couple.

This PR fixes them.

## Why It's Good For The Game

Less bugs, guns will behave as intended, armor will become overpowered (probably)

## Testing

Compiled and ran, test server encouraged, but not essential.

Extended value testing will take place once I'm free. If a test server goes up in the meantime, add this to it.

## Changelog
:cl:
fix: fixed calculation errors in penetration and armor
/:cl:

### CODE REVIEW

**Comment by DiaFRAME444** [comment]
> Compiled and ran, test server encouraged, but not essential.

So you didn't even shoot a few roaches on a test server?

**Comment by Humonitarian** [comment]
> > Compiled and ran, test server encouraged, but not essential.
> 
> So you didn't even shoot a few roaches on a test server?

I refer to public test server

**Comment by Humonitarian** [comment]
> and if you didn't then... lmao...

Do you want to talk about it? I can see there's a problem, but do you want to talk about it?

**Comment by DiaFRAME444** [comment]
Well, I can't actually _DO_ anything about this since I'm not a Maintainer/Project Lead.

**Comment by driecg36** [comment]
Did we not establish with this whole thing that we should test if things work as intended? That's where the issue originally arose and that's how I exposed it, by testing it.

**Comment by Firefox13** [comment]
изображение
By "compiled and run" you mean "I ran the server and compared new values to old values and it seems that with these changes everything works as it should, unlike before" or "It didn't crash"?

**Comment by driecg36** [comment]
@Humonitarian 

I will literally help you test this when I have the time, just anything but another un-tested merge please

File: code/modules/mob/living/living_defense.dm [hunk 1]
```diff
@@ -53,15 +53,15 @@
 			if(dmg_type in list(BRUTE, BURN, TOX, BLAST)) // Some damage types do not help penetrate armor
 				if(remaining_armor)
 					var/dmg_armor_difference = dmg - remaining_armor
-					used_armor += dmg_armor_difference ? dmg - dmg_armor_difference : dmg
-					remaining_armor = dmg_armor_difference ? 0 : -dmg_armor_difference
-					dmg = dmg_armor_difference ? dmg_armor_difference : 0
+					used_armor += dmg_armor_difference > 0 ? dmg - dmg_armor_difference : dmg
+					remaining_armor = dmg_armor_difference > 0 ? 0 : -dmg_armor_difference
+					dmg = dmg_armor_difference > 0 ? dmg_armor_difference : 0
```
Comment (Pink-Chink) [inline code/modules/mob/living/living_defense.dm:58]:
```suggestion
					var/is_difference_positive = dmg_armor_difference > 0
					used_armor += is_difference_positive ? dmg - dmg_armor_difference : dmg
					remaining_armor = is_difference_positive ? 0 : -dmg_armor_difference
					dmg = is_difference_positive ? dmg_armor_difference : 0
```

File: code/modules/mob/living/living_defense.dm [hunk 2]
```diff
@@ -53,15 +53,15 @@
...
 				if(remaining_ablative && dmg)
 					var/ablative_difference
 					ablative_difference = dmg - remaining_ablative
-					used_armor += ablative_difference ? dmg - ablative_difference : dmg
-					remaining_ablative = ablative_difference ? 0 : -ablative_difference
-					dmg = ablative_difference ? ablative_difference : 0
+					used_armor += ablative_difference > 0 ? dmg - ablative_difference : dmg
+					remaining_ablative = ablative_difference > 0 ? 0 : -ablative_difference
+					dmg = ablative_difference > 0 ? ablative_difference : 0
```
Comment (Pink-Chink) [inline code/modules/mob/living/living_defense.dm:64]:
```suggestion
					var/is_difference_positive = ablative_difference > 0
					used_armor += is_difference_positive ? dmg - ablative_difference : dmg
					remaining_ablative = is_difference_positive ? 0 : -ablative_difference
					dmg = is_difference_positive ? ablative_difference : 0
```

File: code/modules/projectiles/projectile.dm [hunk 3]
```diff
@@ -628,19 +628,19 @@
 			dmg_total += dmg
 		if(dmg && amount)
 			var/dmg_armor_difference = dmg - amount
-			amount = dmg_armor_difference ? 0 : -dmg_armor_difference
-			dmg = dmg_armor_difference ? dmg_armor_difference : 0
+			amount = dmg_armor_difference > 0 ? 0 : -dmg_armor_difference
+			dmg = dmg_armor_difference > 0 ? dmg_armor_difference : 0
```
Comment (Pink-Chink) [inline code/modules/projectiles/projectile.dm:632]:
```suggestion
			var/is_difference_positive = dmg_armor_difference > 0
			amount = is_difference_positive ? 0 : -dmg_armor_difference
			dmg = is_difference_positive ? dmg_armor_difference : 0
```

**Comment by Humonitarian** [comment]
This should clarify the intent.

**Comment by Humonitarian** [comment]
Thx for the suggs, looks a lot cleaner now

**Comment by TheShown911** [comment]
Always test ternary statements, very easy to screw something up with them.

File: code/modules/mob/living/living_defense.dm [hunk 4]
```diff
@@ -23,7 +23,7 @@
 			dmg_types += damagetype
 		dmg_types[damagetype] += damage
 
-	if(!armor_divisor)
+	if(armor_divisor <= 0)
 		armor_divisor = 1
 		log_debug("[used_weapon] applied damage to [name] with an armor divisor of 0")
```
Comment (hyperioo) [inline code/modules/mob/living/living_defense.dm:28]:
```suggestion
		log_debug("[used_weapon] applied damage to [name] with a negative or 0 armor divisor")
```

Comment (Humonitarian) [reply code/modules/mob/living/living_defense.dm:28]:
```suggestion
		log_debug("[used_weapon] applied damage to [name] with a nonpositive armor divisor")
```

File: code/modules/projectiles/projectile.dm [hunk 5]
```diff
@@ -628,19 +628,20 @@
...
+			var/is_difference_positive = dmg_armor_difference > 0
+			amount = is_difference_positive ? 0 : -dmg_armor_difference
+			dmg = is_difference_positive ? dmg_armor_difference : 0
 			if(!(dmg_type == HALLOSS))
 				dmg_remaining += dmg
-		if(dmg)
+		if(dmg > 0)
 			damage_types[dmg_type] = dmg
 		else
 			damage_types -= dmg_type
```
Comment (hyperioo) [inline code/modules/projectiles/projectile.dm:639]: No issue with deleting an element of a list you are iterating on?

**Review by hyperioo — COMMENTED** [review]
Let's test this on Monday

**Pull request merged**